File: algorigthms/sorting/insertion_sort.py
def insertion_sort(na):
	''' 关于 insertion sort 要信手拈来 '''
	length = len(na)
	for i in range(1, length):
		j = i
		while j >= 1 and na[j] < na[j-1]:
			tmp = na[j]
			na[j] = na[j - 1]
			na[j - 1] = tmp
			j -= 1

na = [9, 7, 4, 1, 89, 12, 12, 2]
print('Before sorting:', na)
insertion_sort(na)
print('After  sorting:', na)

File: algorigthms/sorting/insertion_sort.cpp
#include <iostream>

using namespace std;

void insertion_sort(int * a, int len)
{
	for(int i = 1; i < len; ++i) {
		int j = i;
		while (j > 0 && a[j] < a[j - 1]) {
			int tmp = a[j];
			a[j] = a[j - 1];
			a[j - 1] = tmp;
			j--;
		}
	}
}

void printData(int * data, int len)
{
	for(int i = 0; i < len; ++i) {
		cout<<data[i] << "\t";
	}
}

int main()
{
	int data[] = {3, 4, 1, 8, 99, 12, 31, 5};
	int len = sizeof(data) / sizeof(data[0]);
	cout<< "Before Sorting:";
	printData(data, len);
	insertion_sort(data, len);
	cout<< "\nAfter  Sorting:";
	printData(data, len);
	return 0;
}